### PR TITLE
rustdoc: use smaller number of colors to distinguish items

### PR DESCRIPTION
**Pull request by jsha**
This reduces visual distractions when reading method signatures.

As discussed in https://github.com/rust-lang/rust/issues/59845#issuecomment-974757191, this categorizes items into one of six colors (down from thirteen):

 - method, function (ochre `#AD7C37`)
 - trait, trait alias (dark slate blue `#6E4FC9`)
 - enum, struct, type alias, union, primitive (maroon `#AD378A`)
 - static, module, keyword, associated type, foreign type (steel blue `#3873AD`)
 - macro (green `#068000`)
 - generic params, self, Self (unmarked black `#000000`)

I slightly tweaked the actual color values so they'd have the same lightness (previously the trait color stood out much more than the others). And I made the color for links in general consistently use steel blue (previously there was a slightly different color for "search-failed").

The ayu and dark themes have been updated according to the same logic. I haven't changed any of the color values in those themes, just their assignment to types.

Demo:

https://rustdoc.crud.net/jsha/fewer-colors/std/string/struct.String.html
https://rustdoc.crud.net/jsha/fewer-colors/std/vec/struct.Vec.html
https://rustdoc.crud.net/jsha/fewer-colors/std/io/trait.Read.html
https://rustdoc.crud.net/jsha/fewer-colors/std/iter/trait.Iterator.html

### CODE REVIEW

**Comment by PatchMixolydic** [comment]
I feel that giving functions and macros the same colour might cause confusion when a macro has the same name as a function. Consider this image from the perspective of a new programmer who wants to find out more about the `panic!` macro:
Image of a rustdoc search showing that `core::panic` and `core:: panicking::panic` have the same colour
Of course, a sufficiently knowledgeable user might already know that `core::panic` is the macro, might know that (most) macros are located in the crate root, or might happen to read the description of `core::panicking::panic`, but this could cause more confusion if there was a function and a macro with the same name in the same module without such helpful documentation. Because of this, I feel that macros should either retain their current green colouration (removing any redundancy if necessary) or be displayed with an exclamation point following their paths (ie. `core::panic!` instead of `core::panic`).

**Comment by GuillaumeGomez** [comment]
I agree with @PatchMixolydic. They're close enough that it would make it confusing.

Another thing I wonder is if we should group by "kind of types". For example (by groups):
 * const, type aliases and trait aliases
 * struct, enum and union
 * trait
 * function
 * macro

I'm not too sure where to put primitives. Currently it's the same color as struct, enum and union, I guess it's fine as is. What do you think?

**Comment by jyn514** [comment]
(ugh, I hate GitHub on mobile)

I think we should only have different colors for things that could be confused in the search page. Since the only things that can have the same name must be in different namespaces, that suggests that we should have one color per namespace (value namespace, type ns, macro ns).

**Comment by jsha** [comment]
Good point about macros, particularly that not all macros are function-like, so styling as functions doesn't necessarily make sense. I've restored macros to having their own color. It was nice having just four colors, but since the primary motivation here is to make method signatures less distracting, and macros don't contribute to that problem, I think that's okay.

> I think we should only have different colors for things that could be confused in the search page.

I think we rely too much on colors in the search page to distinguish types. If distinguishing types is important (and I think it is), we should put the type as text in the search results. That way it's accessible to colorblind people, and also understandable to people who haven't figured out what the colors mean (and it helps learn what the colors mean). But that's a change for another day. :-)

> Since the only things that can have the same name must be in different namespaces, that suggests that we should have one color per namespace (value namespace, type ns, macro ns).

I like in theory the idea of mapping color distinctions onto a language concept (namespaces). This PR actually is pretty close to that. One difference: We still distinguish traits vs (enums, structs, unions), even though [traits are also in the type namespace](https://doc.rust-lang.org/beta/reference/names/namespaces.html). I think the distinction between traits and other types is valuable and I'd like to keep it. Is there a specific change you'd like to make so colors align better with namespaces?

**Comment by jsha** [comment]
(Updated the branch and the PR)

**Comment by camelid** [comment]
When making a large UI change like this, please open a thread on Zulip so the whole team can discuss it :)

**Comment by GuillaumeGomez** [comment]
Looks good to me as is. Thanks for working on this!

> I think we rely too much on colors in the search page to distinguish types. If distinguishing types is important (and I think it is), we should put the type as text in the search results. That way it's accessible to colorblind people, and also understandable to people who haven't figured out what the colors mean (and it helps learn what the colors mean). But that's a change for another day. :-)

This is actually a very good point!

> When making a large UI change like this, please open a thread on Zulip so the whole team can discuss it :)

You're absolutely right, let's hear from others as well.

cc @rust-lang/rustdoc

**Comment by jyn514** [comment]
> cc rust-lang/rustdoc

@GuillaumeGomez please stop doing this, I've asked you to stop :(

**Comment by camelid** [comment]
jsha and I [discussed](https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/use.20smaller.20number.20of.20colors.20to.20distinguish.20items.20.2391480) the trait color on Zulip. I tweaked the trait color to match the "lightness" (see the Zulip thread for details on what that is) of the type color, and it looks more similar to the other colors now.

The current plan is this:

- hyperlinks: `#3873AD`
- value namespace: `#AD7C37`
- type namespace: `#AD378A`
- trait namespace: `#6E4FC9`

**Comment by jsha** [comment]
After some discussion on Zulip I've updated the branch and the demo to lighten the trait color from 5137AD to 6E4FC9. You can see this as a palette at https://color.adobe.com/rustdoc%20proposed%20theme-color-theme-19077250. If you want to play around with it, click "Create using Theme".

**Comment by jyn514** [comment]
@camelid just to clarify - are macros using different colors for function-like and derive macros? Or is there only one color for the macro namespace? I don't know what the current behavior is.

(Either seems fine to me, but seems good to clarify)

File: src/test/rustdoc-gui/anchors.goml [hunk 1]
```diff
@@ -11,7 +11,7 @@ reload:
 
 assert-css: ("#toggle-all-docs", {"color": "rgb(0, 0, 0)"})
 assert-css: (".fqn .in-band a:nth-of-type(1)", {"color": "rgb(0, 0, 0)"})
-assert-css: (".fqn .in-band a:nth-of-type(2)", {"color": "rgb(173, 68, 142)"})
+assert-css: (".fqn .in-band a:nth-of-type(2)", {"color": "rgb(173, 55, 138)"})
```
Comment (camelid) [inline src/test/rustdoc-gui/anchors.goml:14]: Yikes, it's worrying that only one test had to change because of this massive color change ;)

**Comment by jsha** [comment]
There's just one color for all types of macros.

**Comment by camelid** [comment]
> @camelid just to clarify - are macros using different colors for function-like and derive macros? Or is there only one color for the macro namespace? I don't know what the current behavior is.
> 
> (Either seems fine to me, but seems good to clarify)

That's a good point; macros aren't in the color pallete IIRC. @jsha could you put the macro color in the palette instead of the unstable item-info color?

**Comment by jsha** [comment]
> could you put the macro color in the palette instead of the unstable item-info color?

already did, and updated the link above :-) https://color.adobe.com/rustdoc%20proposed%20theme-color-theme-19077288

**Comment by jsha** [comment]
Another small update: I've made associated types the same color as generic type parameters, i.e. unmarked. I've pushed the branch and the demo. See an example at https://rustdoc.crud.net/jsha/fewer-colors/std/iter/trait.Iterator.html#method.intersperse.

**Comment by jsha** [comment]
After some more discussion on Zulip, I've changed the color for associated types back to slate blue (same as links, statics, modules, keywords). I've also updated the demo linked at the top; you may need to hard refresh.

**Comment by camelid** [comment]
> generic params, self, Self, associated type (unmarked black `#000000`)

I think the PR description needs to be updated.

**Comment by jsha** [comment]
PR description has been updated. One issue I noticed: we have a single class `type` that is used for both type aliases and associated types (in some contexts). So for instance in the summary for https://rustdoc.crud.net/jsha/fewer-colors/std/io/trait.Read.html, `Result` is colored slate blue instead of maroon, because it refers to std::io::Result, which is a type alias.

To fix that I need to make a little fix to HTML generation.

File: src/test/rustdoc-gui/anchors.goml (hunk 1 above)
Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/anchors.goml:14]: I think more tests will be added. :)

**Comment by jsha** [comment]
I pushed a fix for an issue where associated types were incorrectly getting generated with `class="type"` (which is meant for typedefs).

I'm happy to write some tests for colors, but it's pretty error prone with browser-ui test as it is today. I filed https://github.com/GuillaumeGomez/browser-UI-test/issues/227 for an update to make it easier to specify expected colors in the same way we specify them in the CSS (hex literals instead of rgba tuples).

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks good to me! I'll wait for the others who commented to take a look first but then r=me.

**Review by jyn514 — APPROVED** [review]

**Comment by jsha** [comment]
@GuillaumeGomez said:

> I'll wait for the others who commented to take a look first but then r=me.

So far we have an approval from @jyn514. @camelid and @PatchMixolydic you were the others who commented. I believe I've satisfactorily addressed all your comments. Good to go?

**Comment by camelid** [comment]
I haven't looked at this in a while, but I think it's fine to land this and make tweaks later.

**Comment by PatchMixolydic** [comment]
Looks good to me :+1:

**Comment by aloucks** [comment]
What is the formal process for requesting this to be all or partially reverted? The PR was a step backwards and universally disliked.

Merging the color options makes things harder to find and understand. I've been looking for green text to identify enums in method signatures and now it's impossible. PR #92660 makes things even worse by combining them all into a single section.

Having distinct colors and sections improves visibility and discovery. The first analogy that comes to mind are phone numbers.

`867-5309` is much easier to visually process than `8675309`.

**Comment by GuillaumeGomez** [comment]
I opened a discussion about it in https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/UI.20regression.3F.

**Comment by Ten0** [comment]
For anyone interested this has bothered me enough that I've created and been testing for a few weeks the following userscript to bring back colors to my rustdoc browsing:
https://gist.github.com/Ten0/564a81a64f9316db4051a4f2afe2b484

(You may use it via Tampermonkey on any major browser)

**Pull request merged**